src/engine/subscription/Subscription.spec.ts

```diff
@@ -0,0 +1,575 @@
+          attributes: ['doesNotExist'],
+        }),
+      ];
+
+      function fn() {
+        processEntitySubscriptions(entity, subscriptions);
+      }
+
+      expect(fn).toThrowErrorMatchingSnapshot();
+    });
+
+    it('should allow for empty attribute lists on DELETE type subscriptions', () => {
+      const subscriptions = [
+        new Subscription({
+          type: SUBSCRIPTION_TYPE_DELETE,
+          name: 'onDrop',
+          description: 'drop item',
+          attributes: [],
+        }),
+      ];
+
+      processEntitySubscriptions(entity, subscriptions);
+    });
+  });
+
+  describe('preProcessor', () => {
+    const testEntity = new Entity({
+      name: 'SomeTestsEntityName',
+      description: 'Just some description',
+      attributes: {
+        someAttribute: {
+          type: DataTypeString,
+          description: 'Just some description',
+          required: false,
+        },
+        anotherAttribute: {
+          type: DataTypeString,
+          description: 'Just some description',
+        },
+      },
+      subscriptions: [
+        new Subscription({
+          name: 'SomeSubWithPreProcessor',
+          type: SUBSCRIPTION_TYPE_CREATE,
+          description: 'build item',
+          attributes: ['someAttribute'],
+          delimiter: '/',
+          // wildCard: '',
+          // pattern: '',
+          preProcessor: (
+            _entity,
+            _id,
+            _source,
+            input,
+            typeName,
+            entitySubscription,
+          ) => {
+            if (entitySubscription.attributes && Object.keys(input).length) {
+              const delimiter = entitySubscription.delimiter;
+              const filled = entitySubscription.attributes
+                .map(attribute => input[attribute])
+                .reduce((acc, curr) => `${acc + delimiter + curr}`, '');
+
+              const topic = `${entitySubscription.name}${_entity.name}${filled}`;
+              return topic;
+            }
+
+            return null;
+          },
+        }),
+      ],
+    });
+
+    const someAttribute = 'test';
+
+    it('should pass through preProcessor if it is declared', async () => {
+      const subscriptionByName = testEntity.getSubscriptionByName(
+        'SomeSubWithPreProcessor',
+      );
+      expect(subscriptionByName).toMatchSnapshot();
+
+      const setup = await generateTestSchema({ entities: [testEntity] });
+      const graphqlSchema = generateGraphQLSchema(setup.configuration);
+      const subscriptionDoc = parse(`
+        subscription someSubWithPreProcessorSomeTestsEntityName($input: SomeSubWithPreProcessorSomeTestsEntityNameInput!) {
+          someSubWithPreProcessorSomeTestsEntityName(input: $input) {
+            someTestsEntityName {
+              # someAttribute
+              anotherAttribute
+            }
+          }
+        }`);
+
+      // subscription topic will be automatically generated by shyft
+      let subscription = (await subscribe({
```

Comment:
We generate the subscribe resolver with the help of subscribe module from graphql, which returns an asyncIterator.